tez-runtime-library/src/main/java/org/apache/tez/dag/library/vertexmanager/VertexManagerWithConcurrentInput.java

```diff
@@ -0,0 +1,234 @@
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.apache.tez.dag.api.EdgeProperty.SchedulingType.CONCURRENT;
+
+public class VertexManagerWithConcurrentInput extends VertexManagerPlugin{
+  private static final Logger LOG = LoggerFactory.getLogger(VertexManagerWithConcurrentInput.class);
+
+  private final Map<String, Boolean> srcVerticesConfigured = Maps.newConcurrentMap();
+  private int managedTasks;
+  private boolean tasksScheduled = false;
+  private AtomicBoolean onVertexStartedDone = new AtomicBoolean(false);
+  private Configuration vertexConfig;
+  private String vertexName;
+  private ConcurrentEdgeTriggerType edgeTriggerType;
+  private boolean allSrcVerticesConfigured;
+
+  int completedUpstreamTasks;
+
+  public VertexManagerWithConcurrentInput(VertexManagerPluginContext context) {
+    super(context);
+  }
+
+  @Override
+  public void initialize() {
+    if (getContext().getUserPayload() == null) {
+      throw new TezUncheckedException("user payload cannot be null for VertexManagerWithConcurrentInput");
+    }
+    managedTasks = getContext().getVertexNumTasks(getContext().getVertexName());
+    Map<String, EdgeProperty> edges = getContext().getInputVertexEdgeProperties();
+    for (Map.Entry<String, EdgeProperty> entry : edges.entrySet()) {
+      if (!CONCURRENT.equals(entry.getValue().getSchedulingType())){
+        throw new TezUncheckedException("All input edges to vertex " + vertexName +
+            "  must be CONCURRENT.");
+      }
+      String srcVertex = entry.getKey();
+      srcVerticesConfigured.put(srcVertex, false);
+      getContext().registerForVertexStateUpdates(srcVertex, EnumSet.of(VertexState.CONFIGURED));
+    }
+
+    try {
+      vertexConfig = TezUtils.createConfFromUserPayload(getContext().getUserPayload());
+    } catch (IOException e) {
+      throw new TezUncheckedException(e);
+    }
+    edgeTriggerType = ConcurrentEdgeTriggerType.valueOf(
+        vertexConfig.get(TezConfiguration.TEZ_CONCURRENT_EDGE_TRIGGER_TYPE,
+            TezConfiguration.TEZ_CONCURRENT_EDGE_TRIGGER_TYPE_DEFAULT));
+    if (!ConcurrentEdgeTriggerType.SOURCE_VERTEX_CONFIGURED.equals(edgeTriggerType)) {
+      // pending TEZ-3999
+      throw new TezUncheckedException("Only support SOURCE_VERTEX_CONFIGURED triggering type for now.");
+    }
+    LOG.info("VertexManagerWithConcurrentInput initialized with edgeTriggerType {}.", edgeTriggerType);
+
+    vertexName = getContext().getVertexName();
+    completedUpstreamTasks = 0;
+  }
+
+  @Override
+  public void onVertexStarted(List<TaskAttemptIdentifier> completions) {
+    onVertexStartedDone.set(true);
+    scheduleTasks();
+  }
+
+  @Override
+  public void onVertexStateUpdated(VertexStateUpdate stateUpdate) {
+    Preconditions.checkArgument(VertexState.CONFIGURED.equals(stateUpdate.getVertexState()),
```

Comment:
Precondition performance may be a consideration here since message is string is composed regardless of boolean condition. Perhaps replace with 3 argument Precondition or evaluate condition and throw IllegalArgumentExcteption manually, constructing the message in the condition block.
Same for other precondition checks.
https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Preconditions.html#checkArgument(boolean,%20java.lang.String,%20java.lang.Object...)

Reply:
done, replace with check-first-and-throw-if-necessary